extensions/amp-form/0.1/amp-form.js

```diff
@@ -107,16 +127,17 @@ export class AmpForm {
       return;
     }
 
-    const shouldValidate = !this.form_.hasAttribute('novalidate');
-    if (shouldValidate &&
-        this.form_.checkValidity && !this.form_.checkValidity()) {
-      e.preventDefault();
-      // TODO(#3776): Use .mutate method when it supports passing state.
-      this.vsync_.run({
-        measure: undefined,
-        mutate: reportValidity,
-      }, {form: this.form_});
-      return;
+    if (this.shouldValidate_) {
+      checkUserValidityOnSubmission_(this.form_);
+      if (this.form_.checkValidity && !this.form_.checkValidity()) {
```

Comment:
I've changed this a bit, passing `propagate` attribute to `checkUserValidity` and returning the `checkValidity` result from it. This allows us to do the `checkValidity` once on the form.